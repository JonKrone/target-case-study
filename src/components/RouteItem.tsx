import React from 'react'
import useFetch from 'use-http'
import { Link } from 'react-router-dom'

interface RouteItemProps {
  item: NextTrip.Route
}

// Could utilize a virtualized list to limit the fetching
export const RouteItem: React.FC<RouteItemProps> = ({
  item: { Description, Route },
}) => {
  // TODO(krone): handle error
  const { error, data = [] } = useFetch<NextTrip.RouteDirection[]>(
    `https://svc.metrotransit.org/NexTrip/Directions/${Route}`,
    { mode: 'cors' },
    []
  )

  return (
    <li style={{ display: 'flex', justifyContent: 'space-between' }}>
      {Description}
      <div style={{ paddingRight: '1rem' }}>
        {data.map(({ Text, Value }) => (
          <Link key={Value} to={`/${Route}/${Value}`}>
            <button style={{ width: '6rem' }}>{capitalize(Text)}</button>
          </Link>
        ))}
      </div>
    </li>
  )
}

const capitalize = (word: string) =>
  word[0].toLocaleUpperCase() + word.slice(1).toLocaleLowerCase()
