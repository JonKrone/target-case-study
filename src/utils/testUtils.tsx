import React, { ReactElement } from 'react'
import { Router } from 'react-router-dom'
import { createMemoryHistory } from 'history'
import { render } from '@testing-library/react'

export const renderWithWrapper = (
  ui: ReactElement,
  {
    route = '/',
    history = createMemoryHistory({ initialEntries: [route] }),
  }: { route?: string; history?: any } = {}
) => {
  const Wrapper: React.FC<{}> = ({ children }) => (
    <Router history={history}>{children}</Router>
  )

  return {
    ...render(ui, { wrapper: Wrapper }),
    // adding `history` to the returned utilities to allow us
    // to reference it in our tests (just try to avoid using
    // this to test implementation details).
    history,
  }
}
