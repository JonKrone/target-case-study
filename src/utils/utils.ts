export const capitalize = (word: string) =>
  word[0].toLocaleUpperCase() + word.slice(1).toLocaleLowerCase()

export const cardinalNumMap: Record<string, string> = {
  1: 'SOUTHBOUND',
  2: 'EASTBOUND',
  3: 'WESTBOUND',
  4: 'NORTHBOUND',
  SOUTHBOUND: '1',
  EASTBOUND: '2',
  WESTBOUND: '3',
  NORTHBOUND: '4',
}
