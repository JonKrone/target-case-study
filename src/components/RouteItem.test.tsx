import React from 'react'
import useFetch from 'use-http'

import { RouteItem } from './RouteItem'
import { renderWithWrapper } from '../utils/testUtils'

jest.mock('use-http', () => ({
  __esModule: true,
  default: jest.fn(() => mockUseFetchData),
}))

const mockUseFetchData = { data: [{ Text: 'Southbound', Value: '1' }] }
const testRoute = { Description: 'Secret', Route: '81', ProviderID: 'James' }

describe('RouteItem', () => {
  it('renders', () => {
    const { asFragment, getByText } = renderWithWrapper(
      <RouteItem item={testRoute} />
    )

    expect(getByText(/Secret/)).toBeTruthy()
    expect(getByText(/Southbound/)).toBeTruthy()

    expect(asFragment()).toMatchSnapshot()
  })

  it('invokes useFetch', () => {
    renderWithWrapper(<RouteItem item={testRoute} />)
    expect(useFetch).toHaveBeenCalledWith(
      expect.objectContaining({ path: '/Directions/81' }),
      []
    )
  })

  it('infers active state from the URL route', () => {
    const { container } = renderWithWrapper(<RouteItem item={testRoute} />, {
      route: '/81/2',
    })

    expect(container.querySelector('li')).toHaveStyle(
      'background-color: #00a896'
    )
  })

  it('clicking a Direction button changes the URL', () => {
    const { history, getByText } = renderWithWrapper(
      <RouteItem item={testRoute} />
    )

    getByText(/Southbound/).click()
    expect(history.location.pathname).toBe('/81/1')
  })
})
