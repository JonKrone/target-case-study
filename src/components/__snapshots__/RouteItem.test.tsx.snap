// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`RouteItem renders 1`] = `
<DocumentFragment>
  <li
    style="display: flex; justify-content: space-between; align-items: center; padding: 1rem 0px;"
  >
    <h4
      style="width: 45%; margin: 0px; padding-left: 1rem;"
    >
      Secret
    </h4>
    <div
      style="padding-right: 1rem;"
    >
      <a
        href="/81/1"
      >
        <button
          style="width: 6rem; padding: 0.5rem;"
          tabindex="-1"
        >
          Southbound
        </button>
      </a>
    </div>
  </li>
</DocumentFragment>
`;
